### DISCUSSION TITLE
使用官方示例web_demo.py时，gradio页面报错，无法显示模型应答结果

### DISCUSSION THREAD

**Discussion by xinaiwunai** [zai-org/ChatGLM3#488] [Bad Case]
![微信图片_20231129231055](https://github.com/THUDM/ChatGLM3/assets/37302121/333b1293-404b-4e8c-ac54-e7194a36c75e)
通过命令启动成功，如下所示：
```cmd
(pytorch) F:\AI\ChatGLM3\basic_demo>python web_demo.py
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:56<00:00,  8.08s/it]
F:\AI\ChatGLM3\basic_demo\web_demo.py:98: GradioDeprecationWarning: The `style` method is deprecated. Please set these arguments in the constructor instead.
  user_input = gr.Textbox(show_label=False, placeholder="Input...", lines=10).style(
Running on local URL:  http://127.0.0.1:8501

To create a public link, set `share=True` in `launch()`.

```
也成功弹出浏览器并进入示例页面，任意输入一句话，提交后。能监控到请求得到了正确的响应，如下图：
![微信图片_20231129231420](https://github.com/THUDM/ChatGLM3/assets/37302121/d9ea02b0-dfa3-40b4-8e13-dbc6b951c808)

但是页面却出现js错误，显示不出应答结果。
具体情况是：大模型响应的第一个字“你”会正常出现，然后试图显示下一个字“好”时就直接报错，同时第一个字“你”也一起消失了。且后续也无法提交新的输入。
浏览器监控console提示如下js错误：
```js
dom.js:238  Uncaught (in promise) DOMException: Failed to execute 'insertBefore' on 'Node': The node before which the new node is to be inserted is not a child of this node.
    at T (http://127.0.0.1:8501/assets/index-a959df42.js:2:12787)
    at Au.i (http://127.0.0.1:8501/assets/index-a959df42.js:2:15665)
    at Au.p (http://127.0.0.1:8501/assets/index-a959df42.js:2:15716)
    at Object.p (http://127.0.0.1:8501/assets/StaticMarkdown-06312fb7.js:339:355)
    at Object.p (http://127.0.0.1:8501/assets/StaticMarkdown-06312fb7.js:339:611)
    at Ba (http://127.0.0.1:8501/assets/index-a959df42.js:5:1914)
    at zo (http://127.0.0.1:8501/assets/index-a959df42.js:5:1548)
```

**Comment by zRzRzRzRzRzRzR** [comment]
确定你的gradio不是4.x版本的么，我们检查一下

**Reply by xinaiwunai** [reply]
gradio是 3.50.2。


(pytorch) F:\AI\ChatGLM3\basic_demo>pip list
Package                   Version
------------------------- ------------
accelerate                0.24.1
aiofiles                  23.2.1
aiohttp                   3.8.6
aiosignal                 1.3.1
altair                    5.2.0
annotated-types           0.6.0
anyio                     3.7.1
argon2-cffi               23.1.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
asttokens                 2.4.1
async-lru                 2.0.4
async-timeout             4.0.3
attrs                     23.1.0
Babel                     2.13.1
beautifulsoup4            4.12.2
bleach                    6.1.0
blinker                   1.7.0
cachetools                5.3.2
certifi                   2023.7.22
cffi                      1.16.0
charset-normalizer        3.3.2
click                     8.1.7
colorama                  0.4.6
comm                      0.2.0
contourpy                 1.2.0
cpm-kernels               1.0.11
cycler                    0.12.1
datasets                  2.14.6
debugpy                   1.8.0
decorator                 5.1.1
defusedxml                0.7.1
dill                      0.3.7
evaluate                  0.4.1
exceptiongroup            1.1.3
executing                 2.0.1
fastapi                   0.104.1
fastjsonschema            2.18.1
ffmpy                     0.3.1
filelock                  3.13.1
fonttools                 4.44.0
fqdn                      1.5.1
frozenlist                1.4.0
fsspec                    2023.10.0
gitdb                     4.0.11
GitPython                 3.1.40
gradio                    3.50.2
gradio_client             0.6.1
h11                       0.14.0
httpcore                  1.0.2
httpx                     0.25.2
huggingface-hub           0.17.3
idna                      3.4
importlib-metadata        6.8.0
importlib-resources       6.1.1
ipykernel                 6.26.0
ipython                   8.17.2
ipywidgets                8.1.1
isoduration               20.11.0
jedi                      0.19.1
Jinja2                    3.1.2
joblib                    1.3.2
json5                     0.9.14
jsonpointer               2.4
jsonschema                4.19.2
jsonschema-specifications 2023.7.1
jupyter                   1.0.0
jupyter_client            8.6.0
jupyter-console           6.6.3
jupyter_core              5.5.0
jupyter-events            0.9.0
jupyter-lsp               2.2.0
jupyter_server            2.10.0
jupyter_server_terminals  0.4.4
jupyterlab                4.0.8
jupyterlab-pygments       0.2.2
jupyterlab_server         2.25.1
jupyterlab-widgets        3.0.9
kiwisolver                1.4.5
latex2mathml              3.76.0
loguru                    0.7.2
Markdown                  3.5.1
markdown-it-py            3.0.0
MarkupSafe                2.1.3
matplotlib                3.8.1
matplotlib-inline         0.1.6
mdtex2html                1.2.0
mdurl                     0.1.2
mistune                   3.0.2
mpmath                    1.3.0
multidict                 6.0.4
multiprocess              0.70.15
nbclient                  0.9.0
nbconvert                 7.11.0
nbformat                  5.9.2
nest-asyncio              1.5.8
networkx                  3.2.1
notebook                  7.0.6
notebook_shim             0.2.3
numpy                     1.26.1
orjson                    3.9.10
overrides                 7.4.0
packaging                 23.2
pandas                    2.1.3
pandocfilters             1.5.0
parso                     0.8.3
Pillow                    10.1.0
pip                       23.3
platformdirs              4.0.0
prometheus-client         0.18.0
prompt-toolkit            3.0.40
protobuf                  4.25.1
psutil                    5.9.6
pure-eval                 0.2.2
pyarrow                   14.0.1
pycparser                 2.21
pydantic                  2.5.2
pydantic_core             2.14.5
pydeck                    0.8.1b0
pydub                     0.25.1
Pygments                  2.16.1
pyparsing                 3.1.1
python-dateutil           2.8.2
python-json-logger        2.0.7
python-multipart          0.0.6
pytz                      2023.3.post1
pywin32                   306
pywinpty                  2.0.12
PyYAML                    6.0.1
pyzmq                     25.1.1
qtconsole                 5.5.0
QtPy                      2.4.1
referencing               0.30.2
regex                     2023.10.3
requests                  2.31.0
responses                 0.18.0
rfc3339-validator         0.1.4
rfc3986-validator         0.1.1
rich                      13.7.0
rpds-py                   0.12.0
safetensors               0.4.0
scikit-learn              1.3.2
scipy                     1.11.3
seaborn                   0.13.0
semantic-version          2.10.0
Send2Trash                1.8.2
sentencepiece             0.1.99
setuptools                68.0.0
six                       1.16.0
smmap                     5.0.1
sniffio                   1.3.0
soupsieve                 2.5
sse-starlette             1.8.2
stack-data                0.6.3
starlette                 0.27.0
streamlit                 1.28.2
sympy                     1.12
tenacity                  8.2.3
terminado                 0.18.0
threadpoolctl             3.2.0
tinycss2                  1.2.1
tokenizers                0.14.1
toml                      0.10.2
tomli                     2.0.1
toolz                     0.12.0
torch                     2.1.0+cu121
torchaudio                2.1.0+cu121
torchvision               0.16.0+cu121
tornado                   6.3.3
tqdm                      4.66.1
traitlets                 5.13.0
transformers              4.35.0
types-python-dateutil     2.8.19.14
typing_extensions         4.8.0
tzdata                    2023.3
tzlocal                   5.2
uri-template              1.3.0
urllib3                   2.0.7
uvicorn                   0.24.0.post1
validators                0.22.0
watchdog                  3.0.0
wcwidth                   0.2.9
webcolors                 1.13
webencodings              0.5.1
websocket-client          1.6.4
websockets                11.0.3
wheel                     0.41.2
widgetsnbextension        4.0.9
win32-setctime            1.1.0
xxhash                    3.4.1
yarl                      1.9.2
zipp                      3.17.0

(pytorch) F:\AI\ChatGLM3\basic_demo>

**Comment by zRzRzRzRzRzRzR** [comment]
![图片](https://github.com/THUDM/ChatGLM3/assets/93239683/99d55b49-5459-4044-b510-3e6663c88707)
你先降级到3.39.0试试

**Reply by xinaiwunai** [reply]
gradio降级3.39.0后，网页正常了。这么说，就是gradio 3.50.2有问题了。
---------------------------------------------------
Successfully installed gradio-3.39.0 linkify-it-py-2.0.2 markdown-it-py-2.2.0 mdit-py-plugins-0.3.3 uc-micro-py-1.0.2
